### ISSUE TITLE
Slow boot after "openssl: 1.0.1t -> 1.0.1u, 1.0.2h -> 1.0.2i, 1.1.0 -> 1.1.0a"

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nixpkgs#18939]
## Issue description
Since the most recent openssl upgrade, "openssl: 1.0.1t -> 1.0.1u, 1.0.2h -> 1.0.2i, 1.1.0 -> 1.1.0a", (on release-16.09: f155746efdbea659e4916a6fbecaa02b6eba8e56, but in reality 2a67ff421ac343be7638f84a2441d2709c4ad136 as ````pythonPackages.cryptography``` doesn't build at the commit that does the upgrade), NixOS boot really slowed down between the "after exiting GRUB menu" and "actually having loaded the kernel" (AFAICT).

The first text seen on the display after exiting the GRUB boot menu is:

```
early console in decompress_kernel
input_data: 0x0000....
input_len: 0x0000...
output: 0x0000...
output_len: 0x0000...
run_size: 0x0000...
KASLR disabled by default...
```

With a pre openssl upgrade boot, this text appears instantly. However, _after_ the upgrade, I can see the addresses being printed one character at a time, overall quite slow. In total, I think the extra wait is about 20 seconds.

See the following videos comparison of boot time:

Normal boot: https://bforsman.name/nextcloud/index.php/s/ObtD4pJd60hOtO4
Slow boot (after openssl upgrade): https://bforsman.name/nextcloud/index.php/s/RvmW2ybMJVA7kZm

Why does this happen? Can we do anything about it? Is it just my hardware?

### Steps to reproduce
Build NixOS 16.09 from 3faa03f87fb9b82b0b5b8b93d5c009a81130d535 as a reference (pre openssl upgrade / normal boot speed). Then build 2a67ff421ac343be7638f84a2441d2709c4ad136 and watch how slow the boot is between exiting the GRUB menu and seeing the ```<<< NixOS Stage 1 >>>``` header.

**Comment by abbradar** [comment]
This seems like a hardware problem to me. However, what do you experience when you try to boot an older generation?

**Comment by bjornfor** [comment]
Older generation boots fast.

**Comment by RonnyPfannschmidt** [comment]
is this an issue only with 16.09? its possible that cryptography silently fails to use openssl and falls back to implementations that are orders of magnitude slower

i had similar troube with borg and had to elevate it by constructing a buildEnv, with headers and libs reunited (which is not nice but worked)

**Comment by dezgeg** [comment]
Weird problem indeed... I would try some "bisection" of this by manually editing the grub config and trying out a couple of combinations like (old kernel, new initrd, new systemConfig), (new kernel, old initrd, old systemConfig) and so on.

**Comment by edolstra** [comment]
I would assume that this is not directly related to the OpenSSL update, as neither GRUB nor the kernel use OpenSSL, as far as I know.

**Comment by edolstra** [comment]
Can you check the size of the initrd? (`stat -L /nix/var/nix/profiles/system/initrd`)

**Comment by abbradar** [comment]
Idea: do your old generations include our hardening changes? Maybe GRUB gets build with some incompatible flag...

**Comment by bjornfor** [comment]
The issue was fixed by 58c9032106c35f7da7040041c70cb2c8b0bc0734 ("linux: 4.4.20 -> 4.4.22").
